src/Database/Expression/WindowExpression.php

```diff
@@ -83,22 +94,102 @@ public function order($fields)
      */
     public function range(?int $start, ?int $end = 0)
     {
-        return $this;
+        if (func_num_args() === 1) {
+            return $this->frame(self::RANGE, $start, self::PRECEDING);
+        } else {
+            return $this->frame(self::RANGE, $start, self::PRECEDING, $end, self::FOLLOWING);
+        }
     }
 
     /**
      * @inheritDoc
      */
     public function rows(?int $start, ?int $end = 0)
     {
-        return $this;
+        if (func_num_args() === 1) {
+            return $this->frame(self::ROWS, $start, self::PRECEDING);
+        } else {
+            return $this->frame(self::ROWS, $start, self::PRECEDING, $end, self::FOLLOWING);
+        }
     }
 
     /**
      * @inheritDoc
      */
     public function groups(?int $start, ?int $end = 0)
     {
+        if (func_num_args() === 1) {
+            return $this->frame(self::GROUPS, $start, self::PRECEDING);
+        } else {
+            return $this->frame(self::GROUPS, $start, self::PRECEDING, $end, self::FOLLOWING);
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function frame(
+        int $type,
+        $startOffset,
+        int $startDirection,
+        $endOffset = null,
+        int $endDirection = self::FOLLOWING
+    ) {
+        $validTypes = [self::RANGE, self::ROWS, self::GROUPS];
+        if (!in_array($type, $validTypes)) {
+            throw new InvalidArgumentException('Frame type must be RANGE, ROWS or GROUP.');
+        }
+
+        $validDirections = [self::PRECEDING, self::FOLLOWING];
+        if (
+            !in_array($startDirection, $validDirections) ||
+            !in_array($endDirection, $validDirections)
+        ) {
+            throw new InvalidArgumentException('Frame directions must be PRECEDING or FOLLOWING.');
+        }
+
+        if (
+            (is_int($startOffset) && $startOffset < 0) ||
+            (is_int($endOffset) && $endOffset < 0)
+        ) {
+            throw new InvalidArgumentException('Frame offsets must be non-negative.');
+        }
+
+        if (
+            in_array($type, [self::ROWS, self::GROUPS]) &&
+            (
+                ($startOffset !== null && !is_int($startOffset)) ||
+                ($endOffset !== null && !is_int($endOffset))
+            )
+        ) {
+            throw new InvalidArgumentException('Frame offsets for ROWS and GROUPS must be null or integers.');
+        }
+
+        if (
+            $type === self::RANGE &&
+            (
+                ($startOffset !== null && !is_int($startOffset) && !is_string($startOffset)) ||
+                ($endOffset !== null && !is_int($endOffset) && !is_string($endOffset))
+            )
+        ) {
+            throw new InvalidArgumentException('Frame offsets for RANGE must be null, integers or interval strings.');
+        }
```

Comment:
What about expression objects? Using functions and other expressions for the frame start/end offset is supported in one way or another, depending on the DBMS.

Reply:
Can you give an specific documentation example that you'd like to see implemented?

Reply:
In MySQL for example there's intervals that can be expressed like `INTERVAL 1 DAY PRECEDING` (**https://dev.mysql.com/doc/refman/8.0/en/window-functions-frames.html**).

The docs don't state it in an overly obvious fashion, but Postgres seems to generally allow expressions that evaluate to a compatible type (with a few limitations), being it function calls `SomeFunction() PRECEDING`, mathematical expressions `(1 + 1) PRECEDING`, or even subqueries `(SELECT foo FROM bar) PRECEDING` (**https://www.postgresql.org/docs/12/sql-expressions.html#SYNTAX-WINDOW-FUNCTIONS**).

Reply:
Getting intervals to work is a goal. Beyond that I'm sure we'll run into issues with non-standard support. I want to avoid supporting too many non-standard implementations.

I don't think we I wrote works for intervals, but it's a starting point to have more than just the integer offset. If adding generic expressions is the solution then that will happen.

Reply:
I also think that accepting `ExpressionInterface` would make a lot of sense here

Reply:
Well mysql only accepts placeholder, integer literals or the interval value. I was hoping to avoid allowing a lot of invalid expressions here. Keep it simple.

Postgres has similar restrictions even if it might allow some expressions.

Reply:
For this iteration, I reduced the allows offsets to integers. Will tackle how to implement intervals when the query tests are in-place so can validate them.